src/response.rs

```diff
@@ -507,18 +562,44 @@ impl Response {
 
         let url = unit.url.clone();
 
-        Ok(Response {
+        // Prefetch data if it's small enough to be buffered in one go.
+        let peeked_len = stream.fill_buf()?.len();
+        let is_fully_buffered = Some(peeked_len) == length;
+
+        let resp = Response {
             url,
             status_line,
             index,
             status,
             headers,
             unit: Box::new(unit),
-            stream: Mutex::new(Box::new(stream.into())),
+            stream: Mutex::new(ResponseStream::Empty),
             history: vec![],
             length,
             compression,
-        })
+        };
+
+        let boxed_stream = ResponseStream::Stream(Box::new(stream.into()));
+
+        let rstream = if is_fully_buffered {
+            // We must read the data through the PoolReturnRead to properly return the stream
+            // to the pool once read to end. This also ensures we handle chunked responses correctly.
```

Comment:
We shouldn't be handling chunked responses here because the content length is known.

Reply:
Very true. Silly me.